### PR TITLE
[Messenger] Move `RejectRedeliveredMessageMiddleware` to AMQP Package

### LINKED ISSUE #41748: [Serializer] Class "Symfony\Component\Messenger\Bridge\Amqp\Transport\AmqpReceivedStamp" does not exist
**Symfony version(s) affected**: 5.3.*, 5.4.*

**Description**
Caused by
ReflectionException: Class "Symfony\Component\Messenger\Bridge\Amqp\Transport\AmqpReceivedStamp" does not exist

**How to reproduce**
Download https://github.com/Warxcell/files and run tests.

**Possible Solution**
Move https://github.com/symfony/symfony/blob/5.3/src/Symfony/Component/Messenger/Middleware/RejectRedeliveredMessageMiddleware.php to https://github.com/symfony/symfony/tree/5.3/src/Symfony/Component/Messenger/Bridge/Amqp namespace, so it will be registered only if AMQP Bridge Package is installed.

### PR DESCRIPTION
**Pull request by Warxcell**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | yes
| Tickets       | Fix #41748
| License       | MIT
| Doc PR        |

### CODE REVIEW

**Comment by xabbuh** [comment]
Can you show the stack trace of the error that you receive? I do not really see why the current code should trigger it.

**Comment by Warxcell** [comment]
> Can you show the stack trace of the error that you receive? I do not really see why the current code should trigger it.

Sure, https://travis-ci.org/github/Warxcell/files/jobs/773978302#L785

If I do `composer require symfony/amqp-messenger` - it's working again. But I don't really need that package.

**Comment by xabbuh** [comment]
I think we should prevent the use of `ReflectionClass` instead if the given class does not exist: #41751

**Comment by Warxcell** [comment]
Why? To me it's looks like this middleware is tighly coupled to AMQP package, why not ship it inside it and load it only when installed, rather than ship it globally in Messenger, load it always, and check on every message if class exists?

**Comment by xabbuh** [comment]
We can think about adding a new middleware in another namespace and deprecate the other one. But such a change must not happen in a patch release and would have to target the `5.4` branch.

**Comment by Warxcell** [comment]
Ok, so I will redirect this to 5.4 and your fix will be patched in 5.2 ?

**Comment by nicolas-grekas** [comment]
Please fix commit+PR title
and add a line in the changelog of the component, and in UPGRADE files for 5.4 and 6.0.

**Comment by kozlice** [comment]
I'm working on an alternative AMQP transport for messenger based on [bunny](https://github.com/jakubkulhan/bunny) library.

Moving this class will complicate dependency tree. Can we choose the way @xabbuh proposes to avoid that?

**EDIT:** This class is indeed so tightly coupled to `symfony/amqp-messenger` it has no business staying at the component level. Please disregard what I said, I'll find a way to make it work for another transport.

**Comment by fabpot** [comment]
@Warxcell Looks good to me. Can you rebase on 6.2 and update the code accordingly? Thank you.

File: UPGRADE-5.4.md
Comment (fabpot) [inline UPGRADE-5.4.md:6]: Missing blank line above

File: UPGRADE-5.4.md
Comment (fabpot) [inline UPGRADE-5.4.md:8]: To be removed

File: UPGRADE-6.0.md
Comment (fabpot) [inline UPGRADE-6.0.md:146]: To be moved to a 6.4 section.

File: UPGRADE-5.4.md
Comment (fabpot) [inline UPGRADE-5.4.md:4]: To be moved to the 6.4 UPGRADE file

**Review by fabpot — CHANGES_REQUESTED** [review]

**Comment by nicolas-grekas** [comment]
On closer look, I think we should close this PR. The linked issue has been fixed, which means that in practice, this is change is just a nice to have. Yes, the implementation knows only about some amqp stamp at the moment, but we could also imagine a day where it manages other kind of stamps.

**Comment by fabpot** [comment]
Thanks for having a look Nicolas. Let's close then.

**Pull request closed without merging**
